File: LeetCodeSolutions/LeetCode_0124.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxPathSum(self, root: TreeNode):
        ret = float('-inf')

        def helper(node):
            nonlocal ret
            if (not node): return 0
            left = max(helper(node.left), 0)
            right = max(helper(node.right), 0)
            ret = max(ret, left + right + node.val)
            return max(right, left) + node.val;

        helper(root)
        return ret

File: LeetCodeSolutions/LeetCode_0124.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    int ret = INT_MIN;
public:
    int maxPathSum(TreeNode* root) {
        helper(root);
        return ret;
    }
    int helper(TreeNode *node){
        if(!node) return 0;
        int left = max(helper(node->left), 0);
        int right = max(helper(node->right),0);
        ret = max(ret, left + right + node->val);
        return max(left, right) + node->val;
    }
};